mistletoe_renderer.py

```diff
@@ -0,0 +1,112 @@
+"""
+This module contains a class CustomHTMLRenderer, which uses
+mistletoe to generate HTML from markdown.
+
+Extra features include:
+- Library note links
+- Managing LaTeX so that MathJax will be able to process it in the browser
+- Syntax highlighting with Pygments
+"""
+import re
+
+from mistletoe import Document, HTMLRenderer, span_token
+from pygments import highlight
+from pygments.lexers import get_lexer_by_name as get_lexer
+from pygments.formatters.html import HtmlFormatter
+
+from mathjax_editing import remove_math, replace_math
+
+
+class NoteLink(span_token.SpanToken):
+    """
+    Detect library note links
+    """
+    parse_inner = False
+    pattern = re.compile(r'Note \[(.*)\]', re.I)
+
+    def __init__(self, match):
+        self.body = match.group(0)
+        self.note = match.group(1)
+
+
+class CustomHTMLRenderer(HTMLRenderer):
+    """
+    Call the constructor with `site_root`.
+
+    The main rendering function is `render_md`.
+    """
+
+    def __init__(self, site_root):
+        self.site_root = site_root
+        super().__init__(NoteLink)
+
+    def render_md(self, ds):
+        """
+        A wrapper for this class's .render() function.
+
+        Input is a string containing markdown with LaTeX,
+        Output is a string containing HTML.
+
+        Uses `mathjax_editing` to strip out sections of the text
+        which potentially contain LaTeX and then splice them back in.
+        """
+        ds_no_math, math = remove_math(ds, '$')
+        # We have to run `mathjax_editing.replace_math` on the text in code
+        # blocks before passing it to Pygments (see `render_block_code`),
+        # otherwise `replace_math` will be confused by the added syntax
+        # highlighting `<span>`s and won't be able to splice in those blocks.
+        self.math = math
+        html = self.render(Document(ds_no_math))
+        return replace_math(html, self.math)
```

Comment:
I don't quite follow why this is necessary - does `mistletoe.latex_token.Math` not work here?

Reply:
I posted a reply [here](https://github.com/leanprover-community/doc-gen/pull/115#issuecomment-779955840). But let's see how #116 turns out!